securedrop/journalist_app/sessions.py

```diff
@@ -0,0 +1,283 @@
+        """This function is called by the flask session interface at the
+        beginning of each request.
+        """
+        is_api = request.path.split("/")[1] == "api"
+
+        if is_api:
+            self.key_prefix = self.api_key_prefix
+            self.salt = self.api_salt
+            auth_header = request.headers.get(self.header_name)
+            if auth_header:
+                split = auth_header.split(" ")
+                if len(split) != 2 or split[0] != "Token":
+                    return self._new_session(is_api)
+                sid: Optional[str] = split[1]
+            else:
+                return self._new_session(is_api)
+        else:
+            sid = request.cookies.get(app.session_cookie_name)
+        if sid:
+            try:
+                sid = self._get_signer(app).loads(sid)
+            except BadSignature:
+                sid = None
+        if not sid:
+            return self._new_session(is_api)
+
+        val = self.redis.get(self.key_prefix + sid)
+        if val is not None:
+            try:
+                data = self.serializer.loads(val.decode("utf-8"))
+                token: str = self._get_signer(app).dumps(sid)  # type: ignore
+                return ServerSideSession(sid=sid, token=token, initial=data)
+            except (JSONDecodeError, NotImplementedError):
+                return self._new_session(is_api)
+        # signed session_id provided in cookie is valid, but the session is not on the server
+        # anymore so maybe here is the code path for a meaningful error message
+        msg = gettext("You have been logged out due to inactivity.")
+        return self._new_session(is_api, initial={"_flashes": [("error", msg)]})
+
+    def save_session(  # type: ignore[override] # noqa
+        self, app: Flask, session: ServerSideSession, response: Response
+    ) -> None:
+        """This is called at the end of each request, just
+        before sending the response.
+        """
+        domain = self.get_cookie_domain(app)
+        path = self.get_cookie_path(app)
+        if session.to_destroy:
+            initial: Dict[str, Any] = {"locale": session.locale}
+            if session.flash:
+                initial["_flashes"] = [session.flash]
+            self.redis.delete(self.key_prefix + session.sid)
+            if not session.is_api:
+                # Instead of deleting the cookie and send a new sid with the next request
+                # create the new session already, so we can pass along messages and locale
+                session = self._new_session(False, initial=initial)
+        expires = self.redis.ttl(name=self.key_prefix + session.sid)
+        if session.new:
+            session["renew_count"] = self.renew_count
+            expires = self.lifetime
+        else:
+            if expires < (30 * 60) and session["renew_count"] > 0:
+                session["renew_count"] -= 1
+                expires += self.lifetime
```

Comment:
Should this be `expires = self.lifetime`? Otherwise a session lifetime can be set up to `7200+1800` sec which is over the limit in the spec.

Reply:
I think that the spec is badly worded and we should keep this behavior so that the overall duration is equal for every session regardless of when the renew happens. It would be easier to debug and maintain in the future, especially considering that the client will renew almost immediately when it is possible.